onewallet/src/main/java/org/kinecosystem/onewallet/OneWalletClient.kt

```diff
@@ -0,0 +1,111 @@
+    private lateinit var uiHandler: Handler
+    private var linkWalletPresenter: LinkWalletPresenter? = null
+    private var kinAccount: KinAccount? = null
+
+    override fun onActivityCreated(activity: Activity, kinAccount: KinAccount, requestCode: Int,
+                                   actionButtonResId: Int, progressBarResId: Int) {
+        uiHandler = Handler(Looper.getMainLooper())
+        val localStore = LocalStore(activity.applicationContext, "ONE_WALLET")
+
+        linkWalletPresenter = LinkWalletPresenter(localStore)
+        linkWalletPresenter?.onAttach(LinkWalletViewHolder(
+                activity.findViewById(actionButtonResId),
+                activity.findViewById(progressBarResId)))
+
+        kinAccount.publicAddress?.let {
+            setupActionListener(activity, it, requestCode)
+        }
+    }
+
+    private fun setupActionListener(activity: Activity, publicAddress: String, requestCode: Int) {
+        linkWalletPresenter?.view?.actionButton?.setOnClickListener {
+            val model = linkWalletPresenter?.oneWalletActionModel
+            model?.let {
+                if (it.isLinkingButton()) {
+                    val transferManager = TransferManager(activity)
+                    transferManager.intentBuilder(ONE_WALLET_APP_ID, ONE_WALLET_LINK_ACTIVITY)
+                            .addParam(EXTRA_APP_ACCOUNT_PUBLIC_ADDRESS, publicAddress)
+                            .addParam(EXTRA_APP_PACKAGE_ID, BuildConfig.APPLICATION_ID)
+                            .addParam(EXTRA_ACTION_TYPE, it.type.toString())
+                            .build()
+                            .start(requestCode)
+
+                } else {
+                    Toast.makeText(activity, "Topup not yet implemented", Toast.LENGTH_LONG).show()
+                }
+            }
+        }
+    }
+
+    override fun onActivityResult(activity: Activity, resultCode: Int, intent: Intent) {
+
+        linkWalletPresenter?.onLinkWalletStarted()
+        val transferManager = TransferManager(activity)
+        transferManager.processResponse(resultCode, intent, object : TransferManager.AccountInfoResponseListener {
+            override fun onCancel() {
+                linkWalletPresenter?.onLinkWalletCancelled()
+            }
+
+            override fun onError(error: String) {
+                Log.e("Linking", "One Wallet (Kinit) was unable to build the linking transaction. Error is: " + error)
+                linkWalletPresenter?.onLinkWalletError(error)
+            }
+
+            override fun onResult(data: String) {
+                processLinkingTransactionResult(data)
+            }
+        })
+    }
+
+
+    override fun onActivityDestroyed() {
+        linkWalletPresenter?.onDetach()
+    }
+
+    private fun processLinkingTransactionResult(data: String) {
+
+        Log.d("Linking", "The linking transaction envelope is: " + data)
+        executorService.execute {
```

Comment:
I'm a little confused, shouldn't the controlled side build the transaction first? with both set options and manage data and send it to kinit?

Reply:
yes it is. processLinkingTransactionResult may not be a good name, but it is called when the kinit activity that builds the transaction envelope finishes and contains that envelope in the data:String parameter.